### PR TITLE
SP-7493: Improve subscription management in meshd application

### PR DESCRIPTION
**Pull request by przemyslawfierek-silvair**

### CODE REVIEW

File: bluetooth_mesh/application.py [hunk 1]
```diff
@@ -27,6 +27,7 @@
 
 import asyncio
 import logging
+from collections import defaultdict
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/application.py:30]: unused?

File: bluetooth_mesh/application.py [hunk 2]
```diff
@@ -729,12 +730,10 @@ async def attach(self, token: Optional[int] = None):
 
         self.addr = await self.node_interface.address()
 
-        configuration = self._convert_config(configuration)
-
-        for element, models_config in configuration.items():
-            for model_id, model_config in models_config.items():
+        for element, models_configs in configuration.items():
+            for model_tuple, model_config in models_configs.items():
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/application.py:734]: `model_tuple` -> `model_id`

File: bluetooth_mesh/application.py [hunk 3]
```diff
@@ -916,7 +894,7 @@ def dev_key_message_received(
                 return
 
     def update_model_configuration(
-        self, model_id: int, configuration: Mapping[str, Any]
+        self, model_id: int, vendor_id: int, configuration: Mapping[str, Any]
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/application.py:897]: Consistency: make `model_id` a `Tuple[Optional[int], int]` (see models/base.py:61) and pass `vendor_id` as the *first* element of that tuple.

File: bluetooth_mesh/application.py [hunk 4]
```diff
@@ -729,12 +730,10 @@ async def attach(self, token: Optional[int] = None):
...
 
-        configuration = self._convert_config(configuration)
-
-        for element, models_config in configuration.items():
-            for model_id, model_config in models_config.items():
+        for element, models_configs in configuration.items():
+            for model_tuple, model_config in models_configs.items():
                 self.elements[element].update_model_configuration(
-                    model_id, model_config
+                    *model_tuple, model_config
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/application.py:736]: Don't unpack the id tuple, see the comment below.

File: bluetooth_mesh/interfaces.py [hunk 5]
```diff
@@ -31,7 +31,9 @@
 
 import asyncio
 import logging
-from typing import Any, List, Mapping, Tuple
+from collections import defaultdict
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/interfaces.py:34]: unused?

File: bluetooth_mesh/interfaces.py [hunk 6]
```diff
@@ -123,7 +159,7 @@ def dev_key_message_received(
     @method(name="UpdateModelConfiguration")
     def update_model_configuration(self, model_id: "q", configuration: "a{sv}"):
         self.element.update_model_configuration(
-            model_id, {key: val.value for key, val in configuration.items()}
+            model_id, *extract_model_config(configuration)
         )
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/interfaces.py:163]: You're trying to be too smart ;P
```
vendor_id, model_config = extract_model_config(configuration)
self.element.update_model_configuration((vendor_id, model_id), model_config)
```

File: bluetooth_mesh/interfaces.py [hunk 7]
```diff
@@ -257,8 +293,23 @@ async def cancel(self) -> None:
 
     async def attach(
         self, app_defined_root: str, token: int
-    ) -> Tuple[str, Mapping[int, List[Tuple[int, Mapping[str, Any]]]]]:
-        return await self._interface.call_attach(app_defined_root, token)
+    ) -> Tuple[str, Dict[int, Dict[Tuple[int, int], Mapping[str, Any]]]]:
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/interfaces.py:296]: `Tuple[Optional[int], int]`, see models/base.py:61

File: bluetooth_mesh/interfaces.py [hunk 8]
```diff
@@ -257,8 +293,23 @@ async def cancel(self) -> None:
...
+        path, configuration = await self._interface.call_attach(app_defined_root, token)
+
+        configuration_dict = defaultdict(dict)
+
+        for element_config_list in configuration:
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/interfaces.py:301]: `for element, element_config in configuration.items()`

File: bluetooth_mesh/interfaces.py [hunk 9]
```diff
@@ -257,8 +293,23 @@ async def cancel(self) -> None:
...
+            assert len(element_config_list) == 2
+            element, element_conf = element_config_list
+
+            for model_config_list in element_conf:
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/interfaces.py:305]: `for model_id, model_config in element_config.items()`

File: bluetooth_mesh/interfaces.py [hunk 10]
```diff
@@ -257,8 +293,23 @@ async def cancel(self) -> None:
...
+                assert len(model_config_list) == 2
+                model_id, model_conf = model_config_list
+                vendor_id, model_conf_dict = extract_model_config(model_conf)
+
+                configuration_dict[element][(vendor_id, model_id)] = model_conf_dict
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/interfaces.py:310]: :+1:

File: bluetooth_mesh/models/base.py [hunk 11]
```diff
@@ -515,3 +518,13 @@ async def bind(self, app_key_index: int) -> "ModelBindStatus":
...
+
+
+class ModelConfig(NamedTuple):
+    """
+    Model Configuration class for mesh models.
+    """
+
+    bindings: List[int] = []
+    publication_period: timedelta = None
+    subscriptions: Set[Union[int, UUID]] = set()
```
Comment (michallowasrzechonek-silvair) [inline bluetooth_mesh/models/base.py:530]: Careful. I think non-scalar default values should be handled using a factory function. Please consult the docs.

```
>>> from typing import NamedTuple
>>> class Foo(NamedTuple):
...   some_list : list = []
... 
>>> foo = Foo()
>>> foo.some_list
[]
>>> foo.some_list.append(1)
>>> bar = Foo()
>>> bar.some_list
[1]
```

File: bluetooth_mesh/interfaces.py (hunk 5 above)
Comment (przemyslawfierek-silvair) [reply bluetooth_mesh/interfaces.py:34]: it is ussed in attach function

File: bluetooth_mesh/interfaces.py (hunk 8 above)
Comment (przemyslawfierek-silvair) [reply bluetooth_mesh/interfaces.py:301]: this is not a dict. this is the list of tuples :/

File: bluetooth_mesh/interfaces.py (hunk 7 above)
Comment (przemyslawfierek-silvair) [reply bluetooth_mesh/interfaces.py:296]: ?

**Pull request merged**